Add support for uploading an Azure image in cloud API

ondrejbudai opened the pull request:
Testing the CI. Will write more later.

This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [x] adequate documentation informing people about the change

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

For user-visible changes, "adequate documentation" is an entry describing the
change for users in docs/news. Please refer to docs/news/README.md for details.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`
-->

teg commented on cmd/osbuild-worker/jobimpl-osbuild.go:
```diff
@@ -282,6 +300,112 @@ func (impl *OSBuildJobImpl) Run(job worker.Job) error {
+
+			var authorizer autorest.Authorizer
+			if impl.AzureCreds == nil {
+				authorizer, err = auth.NewAuthorizerFromEnvironment()
```
The idea is that this would use an Azure Managed Identity if no environment variables are set, right?

ondrejbudai replied:
I've decided to remove this feature for now. With `auth.NewAuthorizerFromEnvironment()`, we cannot specify the tenant ID per job. This should surely be possible but needs more research and it's actually not needed for the MVP.

teg commented on cmd/osbuild-worker/jobimpl-osbuild.go:
For the future: we probably want to have some idea of how long this takes so if it is worth splitting out like what we want to do for AWS. Some logging around the call would do it, I guess.

ondrejbudai replied:
Logging added, fyi it takes about 1 minute to register the image.

teg commented on cmd/osbuild-worker/main.go:
```diff
@@ -154,12 +157,21 @@ func main() {
 			Output:       output,
 			KojiServers:  kojiServers,
 			GCPCredsPath: config.GCP.Credentials,
+			AzureCreds:   azureCredentials,
```
Any reason this is different from GCP? The Azure config is a pointer to a struct, while GCP is not, and the Azure config is parsed in `main()` whilst GCP is passed on as just the path to the creds.

ondrejbudai replied:
I added a comment to this: Basically, the fileformat for Azure credentials is created by us and I want to make sure to fail as soon as possible if it's malformed. We probably want to do something similar for GCP too. If we have wrong credentials, we would like to fail as soon as possible and not wait for the first job.

thozza replied:
Yeah, sounds reasonable to do also for GCP, although it would be just `[]byte`, not our own structure... For GCP, not providing the credentials can still work if the worker is not running on GCP instance with proper roles associated with it. So to exclude the possibility to fail late, we would need to try to authenticate when the worker is starting up (which we discussed with @ondrejbudai). But then again it would probably make sense to be able to configure supported features in worker's configuration, because now every worker would have to be able to authenticate with GCP.

thozza commented on test/cases/api.sh:
```diff
@@ -57,13 +61,21 @@ function checkEnvGCP() {
   "$CLOUD_PROVIDER_GCP")
     checkEnvGCP
     ;;
+  "$CLOUD_PROVIDER_GCP")
```
Copy&Paste error, should be `"$CLOUD_PROVIDER_AZURE"`

ondrejbudai replied:
good catch, thanks!

ondrejbudai commented:
Pushed a new version. I cleaned up the whole PR, added commit messages, cleaned up the `internal/azure/upload` directory and changed my mind on some decisions (see discussion).

The target result for Azure is not implemented here to prevent conflicts with #1262

teg commented:
Very nice!

ondrejbudai commented:
Had to force-push to fix a mockbuild failure.

thozza commented on internal/cloudapi/server.go:
```diff
@@ -205,7 +205,7 @@ func (server *Server) Compose(w http.ResponseWriter, r *http.Request) {
 
 			targets = append(targets, t)
-		} else if uploadRequest.Type == "gcp" {
+		} else if uploadRequest.Type == UploadTypes_gcp {
```
:heart:

thozza commented on internal/target/target.go:
```diff
@@ -71,6 +71,8 @@ func UnmarshalTargetOptions(targetName string, rawOptions json.RawMessage) (Targ
 		options = new(AWSTargetOptions)
 	case "org.osbuild.gcp":
 		options = new(GCPTargetOptions)
+	case "org.osbuild.azure.image":
```
Wouldn't it make sense to have this case right under `org.osbuild.azure`? Not really a big issue...

ondrejbudai replied:
Good point, will move it in some kind of a follow-up.

thozza commented on schutzbot/Jenkinsfile:
```diff
@@ -13,6 +13,10 @@ pipeline {
         GCP_BUCKET = "osbuild-composer-testing"
         GCP_REGION = "us-east4"
+        AZURE_TENANT_ID = "1710d22c-ccf0-4421-8ba7-0135cfaecb90"
+        AZURE_SUBSCRIPTION_ID = "8d026bb1-2a65-454d-a88f-c896db94c4f8"
```
Are any of these IDs secret (not any more anyway)? Just wondering whether someone could use them to upload to our account...

ondrejbudai replied:
Without a client ID and secret, these are just random UUIDs. https://stackoverflow.com/questions/45661109/are-azure-subscription-id-aad-tenant-id-and-aad-app-client-id-considered-secre

teg reviewed (approved):
Still great!

thozza commented on cmd/osbuild-worker/jobimpl-osbuild.go:
```diff
@@ -290,9 +297,122 @@ func (impl *OSBuildJobImpl) Run(job worker.Job) error {
+				r = append(r, err)
+				continue
+			}
+			log.Print("[Azure] 🔑 Logged in Azure")
```
Not an issue, just a comment probably for further discussion. This kind of logging is done in GCP's and AWS's case in the `internal/upload/...` implementation, not in worker's job implementation. In case of GCP, I followed what was already implemented for AWS, but I can see that Azure never had any logging in the internal upload implementation. We should probably decide what we want and should make it consistent for all types of upload targets.

ondrejbudai replied:
Yeah, my thinking is that `internal/upload/...` should serve mainly as a library and thus not log anything... But yeah, we should have a discussion. :)

thozza commented on internal/upload/azure/azurestorage.go:
I implemented a similar check for GCP, but I'm thinking that maybe the uploaded file should be deleted from the storage if the checksum check fails. Not an issue for this set of changes, but I would like to get your point of view anyway.

ondrejbudai replied:
Good point, we should open an issue to track it.

thozza commented on internal/upload/azure/azure.go:
Shouldn't the method's `ctx` argument be used here as well?

ondrejbudai replied:
Fixed in the whole `azure.go` file.

thozza reviewed (commented):
Very nice work!

ondrejbudai commented on test/cases/api.sh:
```diff
@@ -102,6 +114,20 @@ function cleanupGCP() {
+  AZURE_CMD="${AZURE_CMD:-}"
+  if [ -n "$AZURE_CMD" ]; then
+    set +e
+    $AZURE_CMD image delete --resource-group sharing-research --name "$AZURE_IMAGE_NAME"
```
Fixed, I'm now skipping the clean-up when AZURE_IMAGE_NAME is empty.

ondrejbudai commented on test/cases/api.sh:
```diff
@@ -162,13 +191,32 @@ EOM
+function installClientAzure() {
+  if ! hash az; then
+    sudo rpm --import https://packages.microsoft.com/keys/microsoft.asc
+    echo -e "[azure-cli]
```
I've just copy-pasted this from the official docs... I added a link to the docs to the script and left `-e` to prevent confusion as this is just a copy. :)

ondrejbudai commented on test/cases/api.sh:
```diff
@@ -296,13 +344,45 @@ function createReqFileGCP() {
   "$CLOUD_PROVIDER_AWS")
     createReqFileAWS
-    ;;
+  ;;
```
Fixed! Thanks!

ondrejbudai commented on test/cases/api.sh:
```diff
@@ -469,12 +549,30 @@ function verifyInGCP() {
+# Verify image in Azure
+function verifyInAzure() {
+  set +x
+  $AZURE_CMD login --service-principal --username "${AZURE_CLIENT_ID}" --password "${AZURE_CLIENT_SECRET}" --tenant "${AZURE_TENANT_ID}"
```
`set +x` is used to prevent printing of credentials.

thozza replied:
Yeah, I meant confused `set +x` with `set +e`. Sorry for the noise.

ondrejbudai commented on tools/provision.sh:
```diff
@@ -23,12 +23,24 @@ sudo cp -a /usr/share/tests/osbuild-composer/worker/osbuild-worker.toml \
+AZURE_CLIENT_SECRET="${AZURE_CLIENT_SECRET:-}"
+if [[ -n "$AZURE_CLIENT_ID" && -n "$AZURE_CLIENT_SECRET" ]]; then
+    set +x
+    sudo tee /etc/osbuild-worker/azure-credentials.toml > /dev/null << EOF
```
`set +x` is used to prevent printing of credentials.

ondrejbudai commented on cmd/osbuild-worker/jobimpl-osbuild.go:
```diff
@@ -290,9 +297,122 @@ func (impl *OSBuildJobImpl) Run(job worker.Job) error {
 			}
 			// TODO: report back the information below, which is necessary to find and use the image
-			// imageName := t.ImageName
+			// imageName := t.BlobName
```
Fixed by rebase. :-D

ondrejbudai commented:
pushed a new version. I addressed all comments and implemented a simple TargetResult that returns just the image name. We might want to extend it later, this is just a start. :)

teg reviewed (approved):
🥳

ondrejbudai commented:
Tom's changes looks good to me (as an author I cannot approve).

teg reviewed (approved):
Between us, we can approve of each other, I think ;)
